source/opt/passes.cpp

```diff
@@ -71,5 +78,84 @@ bool FreezeSpecConstantValuePass::Process(ir::Module* module) {
   return modified;
 }
 
+bool EliminateDeadConstantPass::Process(ir::Module* module) {
+  analysis::DefUseManager def_use;
+  def_use.AnalyzeDefUse(module);
+  std::unordered_set<ir::Instruction*> working_list;
+  // Traverse all the instructions to get the initial set of dead constants as
+  // working list and count number of real uses for constants. Uses in
+  // annotation instructions do not count.
+  std::unordered_map<ir::Instruction*, size_t> use_counts;
+  std::vector<ir::Instruction*> constants = module->constants();
+  for (auto& c : constants) {
+    uint32_t const_id = c->result_id();
+    size_t count = 0;
+    if (analysis::UseList* uses = def_use.GetUses(const_id)) {
+      for (auto& u : *uses) {
+        if (!ir::IsAnnotationInst(u.inst->opcode()) &&
+            !ir::IsDebugInst(u.inst->opcode())) {
+          count += 1;
+        }
+      }
+    }
+    use_counts[c] = count;
+    if (!count) {
+      working_list.insert(c);
+    }
+  }
+
+  // Start from the constants with 0 uses, back trace through the def-use chain
+  // to find all dead constants.
+  std::unordered_set<ir::Instruction*> dead_consts;
+  while (!working_list.empty()) {
+    ir::Instruction* inst = *working_list.begin();
+    // Back propagate if the instruction contains IDs in its operands.
+    switch (inst->opcode()) {
+      case SpvOp::SpvOpConstantComposite:
+      case SpvOp::SpvOpSpecConstantComposite:
+      case SpvOp::SpvOpSpecConstantOp:
+        for (uint32_t i = 0; i < inst->NumInOperands(); i++) {
+          uint32_t operand_id = inst->GetSingleWordInOperand(i);
+          ir::Instruction* def_inst = def_use.GetDef(operand_id);
+          use_counts[def_inst] -= 1;
+          if (!use_counts[def_inst]) {
+            working_list.insert(def_inst);
+          }
+        }
+        break;
+      default:
+        break;
+    }
+    dead_consts.insert(inst);
+    working_list.erase(inst);
+  }
+
+  // Find all annotation and debug instructions that are referencing dead
+  // constants.
+  std::unordered_set<ir::Instruction*> dead_others;
+  for (auto& dc : dead_consts) {
+    if (analysis::UseList* uses = def_use.GetUses(dc->result_id())) {
+      for (auto& u : *uses) {
+        if (ir::IsAnnotationInst(u.inst->opcode()) ||
+            ir::IsDebugInst(u.inst->opcode())) {
+          dead_others.insert(u.inst);
+        }
+      }
+    }
+  }
+
+  // Turn all dead instructions and uses of them to nop
+  bool modified = false;
+  for (auto& dc : dead_consts) {
+    dc->dbg_line_insts().clear();
```

Comment:
We cannot erase `dbg_line_insts`. Those debug instructions are just physically adjacent (ahead of) to this instruction, but they are not a logical part of this instruction. Turning this instruction to `OpNop` should not affect them. (The weirdness of this issue comes from the flexibility of line-related debug instructions.)